File: py/kem.py
"""IETF X-Wing KEM (Python) — encap/decap matching the Go and Rust ports."""

from __future__ import annotations

from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from kyber_py.ml_kem import ML_KEM_768

from zwing import (
    MLKEM768_CIPHERTEXT_SIZE,
    XWING_CIPHERTEXT_SIZE,
    XWING_SHARED_SIZE,
    X25519_POINT_SIZE,
    combine_xwing,
)
from identity import Identity, IdentityPublic


def xwing_encapsulate(recipient: IdentityPublic) -> tuple[bytes, bytes]:
    """Encapsulate to a recipient. Returns (wire ciphertext, 32-byte shared)."""
    ss_m, ct_m = ML_KEM_768.encaps(recipient.xwing_mlkem_pk)

    eph_sk = x25519.X25519PrivateKey.generate()
    eph_pk = eph_sk.public_key().public_bytes(
        encoding=Encoding.Raw, format=PublicFormat.Raw
    )
    recipient_x25519 = x25519.X25519PublicKey.from_public_bytes(
        recipient.xwing_x25519_pk
    )
    ss_x = eph_sk.exchange(recipient_x25519)

    shared = combine_xwing(ss_m, ss_x, eph_pk, recipient.xwing_x25519_pk)

    wire = bytearray(XWING_CIPHERTEXT_SIZE)
    wire[:MLKEM768_CIPHERTEXT_SIZE] = ct_m
    wire[MLKEM768_CIPHERTEXT_SIZE:] = eph_pk

    return bytes(wire), shared


def xwing_decapsulate(identity: Identity, ciphertext: bytes) -> bytes:
    """Decapsulate using the identity's static X-Wing key."""
    if len(ciphertext) != XWING_CIPHERTEXT_SIZE:
        raise ValueError(
            f"zwing: invalid X-Wing ciphertext size (got {len(ciphertext)})"
        )
    ct_m = ciphertext[:MLKEM768_CIPHERTEXT_SIZE]
    eph_pk = ciphertext[MLKEM768_CIPHERTEXT_SIZE:]

    ss_m = ML_KEM_768.decaps(identity.xwing_mlkem_sk, ct_m)
    eph_pub = x25519.X25519PublicKey.from_public_bytes(eph_pk)
    ss_x = identity.xwing_x25519_sk.exchange(eph_pub)

    return combine_xwing(ss_m, ss_x, eph_pk, identity.xwing_x25519_pk_bytes)

File: js/kem.ts
/**
 * IETF X-Wing KEM — encapsulate / decapsulate built on ML-KEM-768 +
 * X25519 with the SHA3-256 combiner. Wire bytes match the Go and Rust
 * implementations.
 */

import { ml_kem768 } from "@noble/post-quantum/ml-kem.js";
import { x25519 } from "@noble/curves/ed25519.js";

import {
  combineXWing,
  MLKEM768_CIPHERTEXT_SIZE,
  X25519_POINT_SIZE,
  XWING_CIPHERTEXT_SIZE,
  XWING_SHARED_SIZE,
} from "./zwing.js";
import type { Identity, IdentityPublic } from "./identity.js";

/**
 * Encapsulate a fresh X-Wing shared secret to a recipient. Returns the
 * combined wire ciphertext (ML-KEM-768 ct || X25519 ephemeral pk) and
 * the 32-byte shared secret.
 */
export function xwingEncapsulate(recipient: IdentityPublic): {
  ciphertext: Uint8Array;
  shared: Uint8Array;
} {
  const { sharedSecret: ssM, cipherText: ctM } = ml_kem768.encapsulate(
    recipient.xwingMlkemPk,
  );

  const ephSk = crypto.getRandomValues(new Uint8Array(32));
  const ephPk = x25519.getPublicKey(ephSk);
  const ssX = x25519.getSharedSecret(ephSk, recipient.xwingX25519Pk);

  const ssM32: Uint8Array = ssM.slice(0, 32);
  const ssX32: Uint8Array = ssX.slice(0, 32);
  const shared = combineXWing(ssM32, ssX32, ephPk, recipient.xwingX25519Pk);

  const wire = new Uint8Array(XWING_CIPHERTEXT_SIZE);
  wire.set(ctM, 0);
  wire.set(ephPk, MLKEM768_CIPHERTEXT_SIZE);

  // Best-effort wipe of intermediate secrets.
  ephSk.fill(0);
  ssM.fill(0);
  ssX.fill(0);

  return { ciphertext: wire, shared };
}

/**
 * Decapsulate an X-Wing wire ciphertext using `id`'s static keypair.
 * Throws on size mismatch or low-order ephemeral.
 */
export function xwingDecapsulate(id: Identity, ciphertext: Uint8Array): Uint8Array {
  if (ciphertext.length !== XWING_CIPHERTEXT_SIZE) {
    throw new Error(
      `zwing: invalid X-Wing ciphertext size (got ${ciphertext.length})`,
    );
  }
  const ctM = ciphertext.slice(0, MLKEM768_CIPHERTEXT_SIZE);
  const ephPk = ciphertext.slice(MLKEM768_CIPHERTEXT_SIZE);

  const ssM = ml_kem768.decapsulate(ctM, id.xwingMlkemSk);
  const ssX = x25519.getSharedSecret(id.xwingX25519Sk, ephPk);

  const shared = combineXWing(
    ssM.slice(0, 32),
    ssX.slice(0, 32),
    ephPk,
    id.xwingX25519Pk,
  );

  ssM.fill(0);
  ssX.fill(0);
  return shared;
}

export { XWING_CIPHERTEXT_SIZE, XWING_SHARED_SIZE, X25519_POINT_SIZE };
